scripts/redis/scan_redis.py

```diff
@@ -46,26 +45,28 @@ def run():
     client = redis.Redis(host=HOSTNAME, port=PORT, db=DB)
 
     print 'Scanning redis keys'
-    cursor = '0'
     match = None
-
-    log_file = file('redis-stats.log', 'w')
-
     signal.signal(signal.SIGINT, signal_handler)
 
-    while cursor != 0 and loop:
-
-        cursor, data = client.scan(cursor=cursor, match=match)
-
-        for key in data:
-            key_type = client.type(key)
-            size = get_size(client, key, key_type)
-            ttl = client.ttl(key)
-            log_file.write('%s %s %s %d\n' % (key, key_type, str(ttl), size))
-        log_file.flush()
-        time.sleep(DELAY)
-
-    log_file.close()
+    with file('redis-stats.log', 'w') as log_file:
+        for keys in grouper(client.scan_iter(match=match), KEYS_PER_BATCH):
+            if not loop:
+                break
+
+            pipe = client.pipeline()
+            for key in keys:
+                pipe.debug_object(key)
```

Comment:
Should we use this against a production node?
https://redis.io/commands/debug-object
> DEBUG OBJECT is a debugging command that should not be used by clients.

Reply:
I think this warning in redis documentation means that API is unstable and could break between minor version (like fields renamed/removed).

I think it is safe to run against a production node though. It just spits some data about redis internals (including object size) and doesn't do anything dangerous.

Reply:
This still seems like a risky operation to run against production instances or at least master nodes.  For instance in the latest code there is this [loop](
https://github.com/antirez/redis/blob/57786b14e3666db0543b23cadd6ecfd40411a814/src/debug.c#L400) which potentially could iterate across huge linked lists for sets or hashes that have millions of entries.  And the current 2.8 code uses a [trick](https://github.com/antirez/redis/blob/2.8/src/rdb.c#L597) to do a fake save of the object to get its size. All of these operations halt other processing until they complete no matter how large the key it's calculating the size for.

This script prioritized safety over exact size calculations.  Using `debug object` could be a command line option that could be used if you felt it was safe for the particular instance.  The long term plan was to implement `SSCAN`, `HSCAN`, and `ZSCAN` to carefully iterate large keys if we needed to get the true size but just getting the counts work so far to find problem keys.

Reply:
@jkemp101 thanks for digging into performance implications! I agree there's a tradeoff between accuracy and risk.

However I believe just iterating through it in directly memory should be pretty fast. In contrast, doing `sscan`'s and the like will require pushing a lot of data (actual set data is retrieved using scan-family commands) through the network so it will be just dumping almost entire redis data. This feels to me as larger risk as it will be either too slow to be usable (iirc, in current state it took hours to dump entire keyset, and fetching contents would take much longer) or just saturating network IO.

I think we could keep current counting mode but then we should make it very explicit that the metric is very different for strings (character counts which is ≈ size in bytes) than for sets (cardinalities). This actually led me and Jeff the wrong way at first so we spent extra time trying to clear AST cache keys, which were strings. They seemed to represent top keys by that "size" metric but the actual memory was used up in sets/lists.